File: build-connect-mobile/lib/core/services/auth_service.dart
import 'dart:async';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:supabase_flutter/supabase_flutter.dart' hide AuthState;

final authServiceProvider = NotifierProvider<AuthService, AppAuthState>(() {
  return AuthService();
});

class AppAuthState {
  final User? user;
  final String? role;
  final bool isLoading;

  AppAuthState({this.user, this.role, this.isLoading = true});

  AppAuthState copyWith({User? user, String? role, bool? isLoading}) {
    return AppAuthState(
      user: user ?? this.user,
      role: role ?? this.role,
      isLoading: isLoading ?? this.isLoading,
    );
  }
}

class AuthService extends Notifier<AppAuthState> {
  final _supabase = Supabase.instance.client;
  StreamSubscription? _authSubscription;

  @override
  AppAuthState build() {
    _init();
    return AppAuthState(isLoading: true);
  }

  void _init() {
    _authSubscription = _supabase.auth.onAuthStateChange.listen((data) async {
      final session = data.session;
      if (session != null) {
        // Fetch role with retry to avoid race condition with trigger
        final role = await _fetchRoleWithRetry(session.user.id);
        state = AppAuthState(user: session.user, role: role, isLoading: false);
      } else {
        state = AppAuthState(user: null, role: null, isLoading: false);
      }
    });
  }

  /// Retry profile fetch up to 5 times with backoff — matches Angular fix
  Future<String> _fetchRoleWithRetry(String userId, {int maxAttempts = 5, int delayMs = 500}) async {
    for (int attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        final profile = await _supabase
            .from('profiles')
            .select('role')
            .eq('id', userId)
            .maybeSingle();

        if (profile != null && profile['role'] != null) {
          return profile['role'] as String;
        }
      } catch (_) {}

      if (attempt < maxAttempts) {
        await Future.delayed(Duration(milliseconds: delayMs * attempt));
      }
    }
    return 'user'; // fallback
  }

  Future<void> signIn(String email, String password) async {
    await _supabase.auth.signInWithPassword(email: email, password: password);
  }

  Future<void> signUp(String email, String password, String name, String phone, String role) async {
    await _supabase.auth.signUp(
      email: email,
      password: password,
      data: {'full_name': name, 'phone': phone, 'role': role},
    );
  }

  Future<void> signOut() async {
    try { await _supabase.auth.signOut(); } catch (_) {}
    state = AppAuthState(user: null, role: null, isLoading: false);
  }
}

File: build-connect-web/src/app/core/services/auth.service.ts
import { Injectable } from '@angular/core';
import { SupabaseService } from './supabase.service';
import { BehaviorSubject } from 'rxjs';
import { User } from '@supabase/supabase-js';
import { NotificationService } from '../../shared/services/notification.service';

@Injectable({ providedIn: 'root' })
export class AuthService {
    private currentUserSubject = new BehaviorSubject<User | null>(null);
    public currentUser$ = this.currentUserSubject.asObservable();

    private userRoleSubject = new BehaviorSubject<string>('user');
    public userRole$ = this.userRoleSubject.asObservable();

    private profileSubject = new BehaviorSubject<any>(null);
    public profile$ = this.profileSubject.asObservable();

    constructor(
        private supabase: SupabaseService,
        private notificationService: NotificationService
    ) {
        this.supabase.client.auth.onAuthStateChange((_event, session) => {
            const user = session?.user || null;
            this.currentUserSubject.next(user);
            if (user) {
                this.fetchRoleWithRetry(user.id);
                // Setup notifications ONCE per session — prevents duplicate channels
                this.notificationService.setup(user.id);
            } else {
                this.userRoleSubject.next('user');
                this.profileSubject.next(null);
                // Tear down notification channel on logout
                this.notificationService.teardown();
            }
        });
    }

    get currentUser(): User | null {
        return this.currentUserSubject.value;
    }

    get currentRole(): string {
        return this.userRoleSubject.value;
    }

    get currentProfile(): any {
        return this.profileSubject.value;
    }

    /** Retry up to 5 times with backoff — fixes 406 race condition */
    private async fetchRoleWithRetry(userId: string, maxAttempts = 5, delayMs = 500) {
        for (let attempt = 1; attempt <= maxAttempts; attempt++) {
            const { data } = await this.supabase.client
                .from('profiles')
                .select('*')
                .eq('id', userId)
                .maybeSingle();

            if (data && data.role) {
                this.userRoleSubject.next(data.role);
                this.profileSubject.next(data);
                return;
            }

            if (attempt < maxAttempts) {
                await new Promise(resolve => setTimeout(resolve, delayMs * attempt));
            }
        }
        this.userRoleSubject.next('user');
    }

    async logout(): Promise<void> {
        try {
            await this.supabase.client.auth.signOut({ scope: 'local' });
        } catch (e) { console.warn('Logout warning:', e); }
        localStorage.clear();
        sessionStorage.clear();
        window.location.href = '/';
    }

    /** @deprecated Use logout() instead */
    async signOut(): Promise<void> {
        return this.logout();
    }
}
